src/compliment/sources/class_members.clj

```diff
@@ -201,51 +237,71 @@
 
-^{:lite '(defsource :compliment.lite/members :candidates #'members-candidates)}
-(defsource ::members
-  :candidates #'members-candidates
-  :doc #'members-doc)
-
 ;; ## Static members
 
 (defn static-member-symbol?
   "Tests if prefix looks like a static member symbol, returns parsed parts."
   [x]
-  (re-matches #"([^\/\:\.][^\:]*)\/(.*)" x))
+  (when-let [[_klass method :as parts] (qualified-member-symbol? x)]
+    (when-not (str/starts-with? (str method) ".")
+      parts)))
 
-(def static-members-cache
-  "Stores cache of all static members for every class."
+(def ^:private class-members-cache
+  "Members by class"
   (atom {}))
 
-(defn populate-static-members-cache
-  "Populates static members cache for a given class."
+(defn- populate-class-members-cache
+  "Populates qualified methods cache for a given class."
   [^Class class]
-  (swap! static-members-cache assoc class
-         (reduce (fn [cache ^Member c]
-                   (if (static? c)
-                     (update cache (.getName c) (fnil conj []) c)
-                     cache))
-                 {} (concat (.getMethods class) (.getFields class)))))
+  (swap! class-members-cache assoc class
+         (let [methods&fields (concat (.getMethods class) (.getFields class))
+               constructors   (.getConstructors class)
+               member->type   #(cond
+                                 (not (static? %))   :method
+                                 (instance? Field %) :static-field
+                                 (static? %)         :static-method)
+               update-cache   (fn [cache member name type]
+                                (update cache name
+                                        (fnil conj
+                                              (with-meta [] {:type type})) member))]
+           (reduce (fn [cache ^Member m]
+                     (update-cache cache m "new" :constructor))
+                   (reduce (fn [cache ^Member m]
+                             (update-cache cache m (.getName m) (member->type m)))
+                           {} methods&fields) constructors))))
 
 (defn static-members
   "Returns all static members for a given class."
   [^Class class]
-  (or (@static-members-cache class)
-      (get (populate-static-members-cache class) class)))
+  (let [class-members  (or (@class-members-cache class)
+                           (get (populate-class-members-cache class) class))
+        matching-types (cond-> #{:static-field :static-method}
+                         (clojure-1-12+?) (conj :constructor))]
+    (->> class-members
+         (filter (comp matching-types :type meta val))
+         (into {}))))
+
+(defn non-static-members
+  "Returns all non-static members for a given class."
+  [^Class class]
+  (let [class-members (or (@class-members-cache class)
+                          (get (populate-class-members-cache class) class))]
+    (->> class-members
+         (filter (comp #{:method} :type meta val))
+         (into {}))))
```

Comment:
I want this and the previous functions to allocate nothing on the cached path. The cache would have to separate the members at cache construction time. I can't say off the top of my head what's the main criteria for separation should be (static vs non-static, or qualified vs non-qualified). Maybe, these two functions should be dropped completely, and the cache fetching/filtering should be performed in `members-candidates`/`static-members-candidates` directly – that's one of the options.